src/device_update.rs

```diff
@@ -0,0 +1,330 @@
+
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+#[tokio::main]
+pub async fn import_update(
+    import_manifest_path: &Path,
+    container_name: String,
+    tenant_id: String,
+    client_id: String,
+    client_secret: String,
+    instance_id: String,
+    device_update_endpoint_url: &Url,
+    blob_storage_account: String,
+    blob_storage_key: String,
+) -> Result<()> {
+    // credentials for deviceupdate
+    let creds = std::sync::Arc::new(ClientSecretCredential::new(
+        azure_core::new_http_client(),
+        tenant_id,
+        client_id,
+        client_secret,
+        TokenCredentialOptions::default(),
+    ));
+    let client = DeviceUpdateClient::new(device_update_endpoint_url.as_str(), creds)?;
+    let manifest_file_size = std::fs::metadata(import_manifest_path)
+        .context(format!(
+            "cannot get file metadata of {}",
+            import_manifest_path
+                .to_str()
+                .context("import manifest pah invalid")?
+        ))?
+        .len();
+    let manifest_sha256 = base64::encode_config(
+        sha2::Sha256::digest(std::fs::read(import_manifest_path).unwrap()),
+        base64::STANDARD,
+    );
+
+    let manifest: serde_json::Value = serde_json::from_reader(
+        OpenOptions::new()
+            .read(true)
+            .create(false)
+            .open(import_manifest_path)
+            .context("open import manifest file")?,
+    )
+    .context("read import manifest file")?;
+
+    let file_name1 = manifest["instructions"]["steps"][1]["files"][0]
+        .as_str()
+        .context("step1 file not found")?;
+    let file_name2 = manifest["instructions"]["steps"][1]["files"][1]
+        .as_str()
+        .context("step2 file not found")?;
+
+    let storage_credentials =
+        StorageCredentials::access_key(&blob_storage_account, &blob_storage_key);
+    let storage_account_client = BlobServiceClient::new(&blob_storage_account, storage_credentials);
+    let container_client = storage_account_client.container_client(&container_name);
+    let import_manifest_path = import_manifest_path.file_name().unwrap().to_str().unwrap();
+    let manifest_url = generate_sas_url(&container_client, import_manifest_path).await?;
+    let file_url1 = generate_sas_url(&container_client, file_name1).await?;
+    let file_url2 = generate_sas_url(&container_client, file_name2).await?;
+    let import_manifest = json!(
+    [
+        {
+            "importManifest":
+            {
+                "url": manifest_url,
+                "sizeInBytes": manifest_file_size,
+                "hashes": {
+                    "sha256": manifest_sha256
+                }
+            },
+            "files": [
+                {
+                    "filename": file_name1,
+                    "url": file_url1
+                },
+                {
+                    "filename": file_name2,
+                    "url": file_url2
+                }
+            ]
+        }
+    ]);
```

Comment:
Is it to be expected that the manifest format may change in the future? IMHO it would be more sensible to add a #[derive(Deserialize)] above and to access the attributes per struct access/to construct the types as structs. This way changes in the format would reflect in compile time errors at any place in the program.